iXBRLViewerPlugin/xhtmlserialize.py

```diff
@@ -13,29 +13,183 @@
 # limitations under the License.
 
 from lxml import etree
+from enum import Enum
 import re
 
+XHTML_NS = 'http://www.w3.org/1999/xhtml'
+XML_NS = 'http://www.w3.org/XML/1998/namespace'
+
+class EscapeMode(Enum):
+    DEFAULT = 0
+    STYLE = 1
+
 class XHTMLSerializer:
 
     # From https://www.w3.org/TR/html401/index/elements.html
-    selfClosableElements = (
+    SELF_CLOSABLE = (
         'area', 'base', 'basefont', 'br', 'col', 'frame', 'hr', 'img', 
         'input', 'isindex', 'link', 'meta', 'param'
     )
 
-    def _expandEmptyTags(self, xml):
-        """
-        Expand self-closing tags
-
-        Self-closing tags cause problems for XHTML documents when treated as
-        HTML.  Tags that are required to be empty (e.g. <br>) are left as
-        self-closing.
-        """
-        for e in xml.iter('*'):
-            m = re.match(r'\{http://www\.w3\.org/1999/xhtml\}(.*)', e.tag)
-            if (m is None or m.group(1) not in XHTMLSerializer.selfClosableElements) and e.text is None:
-                e.text = ''
-
-    def serialize(self, xmlDocument, fout):
-        self._expandEmptyTags(xmlDocument)
-        fout.write(etree.tostring(xmlDocument, method="xml", encoding="utf-8", xml_declaration=True))
+    ESCAPES = {
+        ']]>': ']]&gt;',
+        '>': '&gt;',
+        '<': '&lt;',
+        '"': '&quot;',
+        '&': '&amp;'
+        }
+
+    MUST_ESCAPE_CHARS = r'<&\u0001-\u0008\u000B\u000C\u000E\u001F\u007F-\u009F'
```

Comment:
Not a blocker. I'm curious about these control chars like delete `\u007F` @paulwarren-wk . Is that included because it can actually delete the previous character in certain situations? I've never seen that happen but I wouldn't be surprised I guess, and from what I could tell the xml spec like https://www.w3.org/TR/xml/#NT-AttValue didn't care about these ranges.

Reply:
7F-9F are included because XML 1.1 [requires that they be escaped](https://www.w3.org/TR/2006/REC-xml11-20060816/#sec-xml11):

> In order to improve the robustness of character encoding detection, the additional control characters #x7F through #x9F, which were freely allowed in XML 1.0 documents, now must also appear only as character references.

It's not clear if XML 1.1 is permitted by iXBRL, but we agreed to try to support it in the serializer.

Reply:
Okay, thanks for the info @paulwarren-wk .